hamilton/experimental/h_async.py

```diff
@@ -95,45 +192,140 @@ class AsyncDriver(driver.Driver):
+                "Please pass in a single result builder"
+            )
+        elif len(result_builders) == 0:
+            result_builders = [base.DictResult()]
+        result_builder = result_builders[0]
         super(AsyncDriver, self).__init__(
-            config, *modules, adapter=AsyncGraphAdapter(result_builder=result_builder)
+            config,
+            *modules,
+            adapter=[
+                # We pass in the async adapters here as this can call node-level hooks
+                # Otherwise we trust the driver/fn graph to call sync adapters
+                AsyncGraphAdapter(
+                    result_builder=result_builder,
+                    async_lifecycle_adapters=lifecycle_base.LifecycleAdapterSet(*async_adapters),
+                ),
+                # We pass in the sync adapters here as this can call
+                *sync_adapters,
+                *async_adapters,  # note async adapters will not be called during synchronous execution -- this is for access later
+            ],
         )
+        self.initialized = False
+
+    async def ainit(self) -> "AsyncDriver":
+        """Initializes the driver when using async. This only exists for backwards compatibility.
+        In Hamilton 2.0, we will be using an asynchronous constructor.
+        See https://dev.to/akarshan/asynchronous-python-magic-how-to-create-awaitable-constructors-with-asyncmixin-18j5.
+        """
+        if self.initialized:
+            # this way it can be called twice
+            return self
+        if self.adapter.does_hook("post_graph_construct", is_async=True):
+            await self.adapter.call_all_lifecycle_hooks_async(
+                "post_graph_construct",
+                graph=self.graph,
+                modules=self.graph_modules,
+                config=self.config,
+            )
+        await self.adapter.ainit()
+        self.initialized = True
+        return self
 
     async def raw_execute(
         self,
         final_vars: typing.List[str],
         overrides: Dict[str, Any] = None,
         display_graph: bool = False,  # don't care
         inputs: Dict[str, Any] = None,
-        run_id: str = None,
+        _fn_graph: graph.FunctionGraph = None,
     ) -> Dict[str, Any]:
         """Executes the graph, returning a dictionary of strings (node keys) to final results.
 
         :param final_vars: Variables to execute (+ upstream)
         :param overrides: Overrides for nodes
         :param display_graph: whether or not to display graph -- this is not supported.
         :param inputs:  Inputs for DAG runtime calculation
+        :param _fn_graph: Function graph for compatibility with superclass -- unused
         :return: A dict of key -> result
         """
+        assert _fn_graph is None, (
+            "_fn_graph must not be provided "
+            "-- the only reason you'd do this is to use materialize(), which is not supported yet.."
+        )
+        run_id = str(uuid.uuid4())
```

Comment:
this should be passed in if we're trying to be consistent.

Reply:
It's consistent? https://github.com/DAGWorks-Inc/hamilton/blob/5585c24c7b3813fcfadf180c8c24771e43781e15/hamilton/driver.py#L642